refund.md

```diff
@@ -38,12 +38,29 @@ Execution gas cost can be split into two subparts:
 As we want to guarantee a refund will always happen, let's iterate over possible reasons for revert that might prevent it:
 * **external execution** - this is revert happening in external to our smart contract execution - when doing an external call (that will be stored as an internal transaction in the blockchain history). The semantics of `call` guarantees that the call will return regardless of revert, so this case is covered.
 * **internal execution** - this revert happens when executing ERC1077 code. This is the most tricky part, as contract might want to do a number of preliminary checks that need to be updated
-* **out of gas exception** - preventing this can be achieved by preminary checks of gas at the beginning of transaction
+* **out of gas exception** - preventing this can be achieved by preminary checks of gas at the beginning of transaction. Another way to protect relayer, from out of gas exception, is to reduce gas for call by amount needed to refund: 
+    ```
+    function execute(...) {
+        uint maxgas = min(gaslimit, gasToken.balanceOf(this)/gasPrice);
+        to.call.gas(maxgas - 50000)(data);
+        // Does not work for modules
+        refund();
+    }
+    ``` 
+
 * **not enough funds** - this happens when operation used more gas then the contract can pay for. This can be prevented by calculating max gas and the start of smart contract execution.
 
 ## Big Data attack
 Big data attack happens when an attacker creates a message with a data field of significant size. The transaction is propagated by relayer but fails on preliminary checks (e.g. not enough gas to process transaction further) Relayer needs to pay for the whole transaction, but a refund will not happen. An attacker can drain relayer wallet that way. This can not be prevented in the general case.
 
+Our proposal: 
+```
+ function execute(...) {
+    require(msg.sender == allowedRelayer); 
+ }
+```
+Only selected relayer is allowed to depute execution.
```

Comment:
Move one section down
Our proposal -> Example
Add note that `require` can be implemented in the module
relayer -> relayer(s)